File: 2222-number-of-ways-to-select-buildings/2222-number-of-ways-to-select-buildings.py
class Solution:
    def numberOfWays(self, s: str) -> int:
        pattern1 = "101"
        pattern2 = "010"
        dp1 = [0] * 4
        dp2 = [0] * 4
        dp1[3] = 1
        dp2[3] = 1
        
        for c in s:
            for j in range(3):
                if c == pattern1[j]:
                    dp1[j] += dp1[j+1]
                if c == pattern2[j]:
                    dp2[j] += dp2[j+1]
        return dp1[0] + dp2[0]
        

File: 2222-number-of-ways-to-select-buildings/2222-number-of-ways-to-select-buildings.js
/**
 * @param {string} s
 * @return {number}
 */
var numberOfWays = function(s) {
    const p1 = "101"
    const p2 = "010"
    const dp1 = Array(4).fill(0)
    const dp2 = Array(4).fill(0)
    dp1[3] = dp2[3] = 1
    
    for (let c of s)
        for(let i = 0; i< 3; i++){
            if (p1[i] == c)
                dp1[i] += dp1[i+1]
            if (p2[i] == c)
                dp2[i] += dp2[i+1]
        }
    return dp1[0] + dp2[0]
    
};